inst/tests/mergelist.Rraw

```diff
@@ -114,6 +128,251 @@ test(13.4, cbindlist(list(data.table(a=1:2), data.table(b=3:4, key="b"))), data.
 # TODO(#7116): this could be supported
 # test(13.5, cbindlist(list(data.table(a=1:2, key="a"), data.table(b=3:4, key="b"))), data.table(a=1:2, b=3:4, key=c("a", "b")))
 
+# mergepair
+
+## test copy-ness argument in mergepair
+
+### LHS equal to RHS: no copy in all cases
+local({
+  l = list(
+    lhs = data.table(id1=1:2, v1=1:2),
+    rhs = data.table(id1=1:2, v2=1:2)
+  )
+  expected = data.table(id1=1:2, v1=1:2, v2=1:2)
+  frac = 0.0
+  for (how in c("inner", "left", "right", "full")) {
+    frac = frac + 0.1
+    for (mult in c("all", "first", "last", "error")) {
+      frac = frac + 0.01
+      test(21 + (frac <- frac + 0.001), ans <- mergepair(l$lhs, l$rhs, on="id1", how=how, mult=mult, copy=TRUE), expected) ## copy=TRUE: no shared columns
+      test(21 + (frac <- frac + 0.001), copied(ans, l))
+      test(21 + (frac <- frac + 0.001), ans <- mergepair(l$lhs, l$rhs, on="id1", how=how, mult=mult, copy=FALSE), expected) ## copy=FALSE: LHS shared but no RHS
+      test(21 + (frac <- frac + 0.001), notcopied(ans, l, how=how))
+    }
+  }
+})
+
+### RHS includes LHS: no copy in inner, left, right
+local({
+  l = list(
+    lhs = data.table(id1=1:2, v1=1:2),
+    rhs = data.table(id1=1:3, v2=1:3)
+  )
+  expected = list(
+    inner = data.table(id1=1:2, v1=1:2, v2=1:2),
+    left = data.table(id1=1:2, v1=1:2, v2=1:2),
+    right = data.table(id1=1:3, v1=c(1:2, NA), v2=1:3),
+    full = data.table(id1=1:3, v1=c(1:2, NA), v2=1:3)
+  )
+  frac = 0.0
+  for (how in c("inner", "left", "right", "full")) {
+    frac = frac + 0.1
+    for (mult in c("all", "first", "last", "error")) {
+      frac = frac + 0.01
+      test(22 + (frac <- frac + 0.001), ans <- mergepair(l$lhs, l$rhs, on="id1", how=how, mult=mult, copy=TRUE), expected[[how]])
+      test(22 + (frac <- frac + 0.001), copied(ans, l))
+      test(22 + (frac <- frac + 0.001), ans <- mergepair(l$lhs, l$rhs, on="id1", how=how, mult=mult, copy=FALSE), expected[[how]])
+      if (how == "full") {
+        test(22 + (frac <- frac + 0.001), copied(ans, l))
```

Comment:
Strictly speaking, IINM, we don't need a copy here -- it's an implementation artifact driving the need to do so.